src/System.Text.ValueBuilder/System/Variant.cs

```diff
@@ -0,0 +1,444 @@
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System
+{
+    /// <summary>
+    /// <see cref="Variant"/> is a wrapper that avoids boxing common value types.
+    /// </summary>
+    public readonly struct Variant
+    {
+        private readonly Union _union;
+        public readonly VariantType Type;
+        private readonly object _object;
+
+        /// <summary>
+        /// Get the value as an object if the value is stored as an object.
+        /// </summary>
+        /// <param name="value">The value, if an object, or null.</param>
+        /// <returns>True if the value is actually an object.</returns>
+        public bool TryGetValue(out object value)
+        {
+            bool isObject = Type == VariantType.Object;
+            value = isObject ? _object : null;
+            return isObject;
+        }
+
+        /// <summary>
+        /// Get the value as the requested type <typeparamref name="T"/> if actually stored as that type.
+        /// </summary>
+        /// <param name="value">The value if stored as (T), or default.</param>
+        /// <returns>True if the <see cref="Variant"/> is of the requested type.</returns>
+        public unsafe bool TryGetValue<T>(out T value) where T : unmanaged
+        {
+            value = default;
+            bool success = false;
+
+            // Checking the type gets all of the non-relevant compares elided by the JIT
+            if((typeof(T) == typeof(bool) && Type == VariantType.Boolean)
+                || (typeof(T) == typeof(byte) && Type == VariantType.Byte)
+                || (typeof(T) == typeof(char) && Type == VariantType.Char)
+                || (typeof(T) == typeof(DateTime) && Type == VariantType.DateTime)
+                || (typeof(T) == typeof(DateTimeOffset) && Type == VariantType.DateTimeOffset)
+                || (typeof(T) == typeof(decimal) && Type == VariantType.Decimal)
+                || (typeof(T) == typeof(double) && Type == VariantType.Double)
+                || (typeof(T) == typeof(Guid) && Type == VariantType.Guid)
+                || (typeof(T) == typeof(short) && Type == VariantType.Int16)
+                || (typeof(T) == typeof(int) && Type == VariantType.Int32)
+                || (typeof(T) == typeof(long) && Type == VariantType.Int64)
+                || (typeof(T) == typeof(sbyte) && Type == VariantType.SByte)
+                || (typeof(T) == typeof(float) && Type == VariantType.Single)
+                || (typeof(T) == typeof(TimeSpan) && Type == VariantType.TimeSpan)
+                || (typeof(T) == typeof(ushort) && Type == VariantType.UInt16)
+                || (typeof(T) == typeof(uint) && Type == VariantType.UInt32)
+                || (typeof(T) == typeof(ulong) && Type == VariantType.UInt64))
+            {
+                // The JIT is able to generate more efficient code when including the
+                // code for CastTo<T>() directly.
+                fixed (void* u = &_union)
```

Comment:
Can you link to or open an issue in CoreCLR for this? I'd like to understand why.

Reply:
https://github.com/dotnet/coreclr/issues/21152